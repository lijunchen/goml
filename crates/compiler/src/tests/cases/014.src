fn test_nested_match(x: (Bool, Bool), y: (Bool, Bool)) {
  match x {
    (true, false) => 
      match y {
        (false, true) => string_println("case1"),
        _ => string_println("case2"),
      },
    _ => 
      match y {
        (false, true) => string_println("case3"),
        _ => string_println("case4"),
      },
  }
}



fn main() {
  let _ = test_nested_match((true, false), (false, true)) in
  let _ = test_nested_match((true, false), (true, false)) in
  let _ = test_nested_match((false, true), (false, true)) in
  let _ = test_nested_match((false, true), (true, false)) in
  ()
}
