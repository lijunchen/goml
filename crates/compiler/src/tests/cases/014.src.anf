fn test_nested_match(x/0: (Bool, Bool), y/1: (Bool, Bool)) -> Unit {
  let x0 = x/0.0 in
  let x1 = x/0.1 in
  match x1 {
    true => {
      let x2 = y/1.0 in
      let x3 = y/1.1 in
      match x3 {
        true => {
          match x2 {
            true => {
              println("case4")
            },
            false => {
              println("case3")
            },
          }
        },
        false => {
          println("case4")
        },
      }
    },
    false => {
      match x0 {
        true => {
          let x4 = y/1.0 in
          let x5 = y/1.1 in
          match x5 {
            true => {
              match x4 {
                true => {
                  println("case2")
                },
                false => {
                  println("case1")
                },
              }
            },
            false => {
              println("case2")
            },
          }
        },
        false => {
          let x6 = y/1.0 in
          let x7 = y/1.1 in
          match x7 {
            true => {
              match x6 {
                true => {
                  println("case4")
                },
                false => {
                  println("case3")
                },
              }
            },
            false => {
              println("case4")
            },
          }
        },
      }
    },
  }
}

fn main() -> Unit {
  let t12 = (true, false) in
  let t13 = (false, true) in
  let mtmp8 = test_nested_match(t12, t13) in
  let t14 = (true, false) in
  let t15 = (true, false) in
  let mtmp9 = test_nested_match(t14, t15) in
  let t16 = (false, true) in
  let t17 = (false, true) in
  let mtmp10 = test_nested_match(t16, t17) in
  let t18 = (false, true) in
  let t19 = (true, false) in
  let mtmp11 = test_nested_match(t18, t19) in
  ()
}