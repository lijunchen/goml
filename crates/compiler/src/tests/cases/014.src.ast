fn test_nested_match(x:(Bool, Bool), y:(Bool, Bool)) {
    match x {
        (true, false) => match y {
              (false, true) => println("case1"),
              _ => println("case2"),
          },
        _ => match y {
              (false, true) => println("case3"),
              _ => println("case4"),
          },
    }
}

fn main() {
    let _ = test_nested_match((true, false), (false, true)) in
    let _ = test_nested_match((true, false), (true, false)) in
    let _ = test_nested_match((false, true), (false, true)) in
    let _ = test_nested_match((false, true), (true, false)) in
    ()
}

