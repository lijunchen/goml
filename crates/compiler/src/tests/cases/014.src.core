fn test_nested_match(x/0: (Bool, Bool), y/1: (Bool, Bool)) -> Unit {
  let x0 = x/0.0 in
  let x1 = x/0.1 in
  match x1 {
    true => {
      let x2 = y/1.0 in
      let x3 = y/1.1 in
      match x3 {
        true => {
          match x2 {
            true => {
              println("case4")
            },
            false => {
              println("case3")
            },
          }
        },
        false => {
          println("case4")
        },
      }
    },
    false => {
      match x0 {
        true => {
          let x4 = y/1.0 in
          let x5 = y/1.1 in
          match x5 {
            true => {
              match x4 {
                true => {
                  println("case2")
                },
                false => {
                  println("case1")
                },
              }
            },
            false => {
              println("case2")
            },
          }
        },
        false => {
          let x6 = y/1.0 in
          let x7 = y/1.1 in
          match x7 {
            true => {
              match x6 {
                true => {
                  println("case4")
                },
                false => {
                  println("case3")
                },
              }
            },
            false => {
              println("case4")
            },
          }
        },
      }
    },
  }
}

fn main() -> Unit {
  let mtmp8 = test_nested_match((true,false), (false,true)) in
  let mtmp9 = test_nested_match((true,false), (true,false)) in
  let mtmp10 = test_nested_match((false,true), (false,true)) in
  let mtmp11 = test_nested_match((false,true), (true,false)) in
  ()
}