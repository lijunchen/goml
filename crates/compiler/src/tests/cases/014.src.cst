FILE@0..599
  FN@0..336
    FnKeyword@0..2 "fn"
    Whitespace@2..3 " "
    Lident@3..20 "test_nested_match"
    PARAM_LIST@20..55
      LParen@20..21 "("
      PARAM@21..38
        Lident@21..22 "x"
        Colon@22..23 ":"
        Whitespace@23..24 " "
        TYPE_TUPLE@24..36
          TYPE_LIST@24..36
            LParen@24..25 "("
            TYPE_BOOL@25..29
              BoolKeyword@25..29 "Bool"
            Comma@29..30 ","
            Whitespace@30..31 " "
            TYPE_BOOL@31..35
              BoolKeyword@31..35 "Bool"
            RParen@35..36 ")"
        Comma@36..37 ","
        Whitespace@37..38 " "
      PARAM@38..53
        Lident@38..39 "y"
        Colon@39..40 ":"
        Whitespace@40..41 " "
        TYPE_TUPLE@41..53
          TYPE_LIST@41..53
            LParen@41..42 "("
            TYPE_BOOL@42..46
              BoolKeyword@42..46 "Bool"
            Comma@46..47 ","
            Whitespace@47..48 " "
            TYPE_BOOL@48..52
              BoolKeyword@48..52 "Bool"
            RParen@52..53 ")"
      RParen@53..54 ")"
      Whitespace@54..55 " "
    BLOCK@55..336
      LBrace@55..56 "{"
      Whitespace@56..59 "\n  "
      EXPR_MATCH@59..331
        MatchKeyword@59..64 "match"
        Whitespace@64..65 " "
        EXPR_LIDENT@65..67
          Lident@65..66 "x"
          Whitespace@66..67 " "
        MATCH_ARM_LIST@67..331
          LBrace@67..68 "{"
          Whitespace@68..73 "\n    "
          MATCH_ARM@73..202
            PATTERN_TUPLE@73..87
              LParen@73..74 "("
              PATTERN_BOOL@74..78
                TrueKeyword@74..78 "true"
              Comma@78..79 ","
              Whitespace@79..80 " "
              PATTERN_BOOL@80..85
                FalseKeyword@80..85 "false"
              RParen@85..86 ")"
              Whitespace@86..87 " "
            FatArrow@87..89 "=>"
            Whitespace@89..97 " \n      "
            EXPR_MATCH@97..202
              MatchKeyword@97..102 "match"
              Whitespace@102..103 " "
              EXPR_LIDENT@103..105
                Lident@103..104 "y"
                Whitespace@104..105 " "
              MATCH_ARM_LIST@105..202
                LBrace@105..106 "{"
                Whitespace@106..115 "\n        "
                MATCH_ARM@115..155
                  PATTERN_TUPLE@115..129
                    LParen@115..116 "("
                    PATTERN_BOOL@116..121
                      FalseKeyword@116..121 "false"
                    Comma@121..122 ","
                    Whitespace@122..123 " "
                    PATTERN_BOOL@123..127
                      TrueKeyword@123..127 "true"
                    RParen@127..128 ")"
                    Whitespace@128..129 " "
                  FatArrow@129..131 "=>"
                  Whitespace@131..132 " "
                  EXPR_CALL@132..155
                    EXPR_LIDENT@132..146
                      Lident@132..146 "string_println"
                    ARG_LIST@146..155
                      LParen@146..147 "("
                      ARG@147..154
                        EXPR_STR@147..154
                          Str@147..154 "\"case1\""
                      RParen@154..155 ")"
                Comma@155..156 ","
                Whitespace@156..165 "\n        "
                MATCH_ARM@165..193
                  PATTERN_WILDCARD@165..167
                    WildcardKeyword@165..166 "_"
                    Whitespace@166..167 " "
                  FatArrow@167..169 "=>"
                  Whitespace@169..170 " "
                  EXPR_CALL@170..193
                    EXPR_LIDENT@170..184
                      Lident@170..184 "string_println"
                    ARG_LIST@184..193
                      LParen@184..185 "("
                      ARG@185..192
                        EXPR_STR@185..192
                          Str@185..192 "\"case2\""
                      RParen@192..193 ")"
                Comma@193..194 ","
                Whitespace@194..201 "\n      "
                RBrace@201..202 "}"
          Comma@202..203 ","
          Whitespace@203..208 "\n    "
          MATCH_ARM@208..325
            PATTERN_WILDCARD@208..210
              WildcardKeyword@208..209 "_"
              Whitespace@209..210 " "
            FatArrow@210..212 "=>"
            Whitespace@212..220 " \n      "
            EXPR_MATCH@220..325
              MatchKeyword@220..225 "match"
              Whitespace@225..226 " "
              EXPR_LIDENT@226..228
                Lident@226..227 "y"
                Whitespace@227..228 " "
              MATCH_ARM_LIST@228..325
                LBrace@228..229 "{"
                Whitespace@229..238 "\n        "
                MATCH_ARM@238..278
                  PATTERN_TUPLE@238..252
                    LParen@238..239 "("
                    PATTERN_BOOL@239..244
                      FalseKeyword@239..244 "false"
                    Comma@244..245 ","
                    Whitespace@245..246 " "
                    PATTERN_BOOL@246..250
                      TrueKeyword@246..250 "true"
                    RParen@250..251 ")"
                    Whitespace@251..252 " "
                  FatArrow@252..254 "=>"
                  Whitespace@254..255 " "
                  EXPR_CALL@255..278
                    EXPR_LIDENT@255..269
                      Lident@255..269 "string_println"
                    ARG_LIST@269..278
                      LParen@269..270 "("
                      ARG@270..277
                        EXPR_STR@270..277
                          Str@270..277 "\"case3\""
                      RParen@277..278 ")"
                Comma@278..279 ","
                Whitespace@279..288 "\n        "
                MATCH_ARM@288..316
                  PATTERN_WILDCARD@288..290
                    WildcardKeyword@288..289 "_"
                    Whitespace@289..290 " "
                  FatArrow@290..292 "=>"
                  Whitespace@292..293 " "
                  EXPR_CALL@293..316
                    EXPR_LIDENT@293..307
                      Lident@293..307 "string_println"
                    ARG_LIST@307..316
                      LParen@307..308 "("
                      ARG@308..315
                        EXPR_STR@308..315
                          Str@308..315 "\"case4\""
                      RParen@315..316 ")"
                Comma@316..317 ","
                Whitespace@317..324 "\n      "
                RBrace@324..325 "}"
          Comma@325..326 ","
          Whitespace@326..329 "\n  "
          RBrace@329..330 "}"
          Whitespace@330..331 "\n"
      RBrace@331..332 "}"
      Whitespace@332..336 "\n\n\n\n"
  FN@336..599
    FnKeyword@336..338 "fn"
    Whitespace@338..339 " "
    Lident@339..343 "main"
    PARAM_LIST@343..346
      LParen@343..344 "("
      RParen@344..345 ")"
      Whitespace@345..346 " "
    BLOCK@346..599
      LBrace@346..347 "{"
      Whitespace@347..350 "\n  "
      EXPR_LET@350..597
        LetKeyword@350..353 "let"
        Whitespace@353..354 " "
        PATTERN_WILDCARD@354..356
          WildcardKeyword@354..355 "_"
          Whitespace@355..356 " "
        Eq@356..357 "="
        Whitespace@357..358 " "
        EXPR_LET_VALUE@358..406
          EXPR_CALL@358..406
            EXPR_LIDENT@358..375
              Lident@358..375 "test_nested_match"
            ARG_LIST@375..406
              LParen@375..376 "("
              ARG@376..391
                EXPR_TUPLE@376..389
                  LParen@376..377 "("
                  EXPR_BOOL@377..381
                    TrueKeyword@377..381 "true"
                  Comma@381..382 ","
                  Whitespace@382..383 " "
                  EXPR_BOOL@383..388
                    FalseKeyword@383..388 "false"
                  RParen@388..389 ")"
                Comma@389..390 ","
                Whitespace@390..391 " "
              ARG@391..404
                EXPR_TUPLE@391..404
                  LParen@391..392 "("
                  EXPR_BOOL@392..397
                    FalseKeyword@392..397 "false"
                  Comma@397..398 ","
                  Whitespace@398..399 " "
                  EXPR_BOOL@399..403
                    TrueKeyword@399..403 "true"
                  RParen@403..404 ")"
              RParen@404..405 ")"
              Whitespace@405..406 " "
        InKeyword@406..408 "in"
        Whitespace@408..411 "\n  "
        EXPR_LET_BODY@411..597
          EXPR_LET@411..597
            LetKeyword@411..414 "let"
            Whitespace@414..415 " "
            PATTERN_WILDCARD@415..417
              WildcardKeyword@415..416 "_"
              Whitespace@416..417 " "
            Eq@417..418 "="
            Whitespace@418..419 " "
            EXPR_LET_VALUE@419..467
              EXPR_CALL@419..467
                EXPR_LIDENT@419..436
                  Lident@419..436 "test_nested_match"
                ARG_LIST@436..467
                  LParen@436..437 "("
                  ARG@437..452
                    EXPR_TUPLE@437..450
                      LParen@437..438 "("
                      EXPR_BOOL@438..442
                        TrueKeyword@438..442 "true"
                      Comma@442..443 ","
                      Whitespace@443..444 " "
                      EXPR_BOOL@444..449
                        FalseKeyword@444..449 "false"
                      RParen@449..450 ")"
                    Comma@450..451 ","
                    Whitespace@451..452 " "
                  ARG@452..465
                    EXPR_TUPLE@452..465
                      LParen@452..453 "("
                      EXPR_BOOL@453..457
                        TrueKeyword@453..457 "true"
                      Comma@457..458 ","
                      Whitespace@458..459 " "
                      EXPR_BOOL@459..464
                        FalseKeyword@459..464 "false"
                      RParen@464..465 ")"
                  RParen@465..466 ")"
                  Whitespace@466..467 " "
            InKeyword@467..469 "in"
            Whitespace@469..472 "\n  "
            EXPR_LET_BODY@472..597
              EXPR_LET@472..597
                LetKeyword@472..475 "let"
                Whitespace@475..476 " "
                PATTERN_WILDCARD@476..478
                  WildcardKeyword@476..477 "_"
                  Whitespace@477..478 " "
                Eq@478..479 "="
                Whitespace@479..480 " "
                EXPR_LET_VALUE@480..528
                  EXPR_CALL@480..528
                    EXPR_LIDENT@480..497
                      Lident@480..497 "test_nested_match"
                    ARG_LIST@497..528
                      LParen@497..498 "("
                      ARG@498..513
                        EXPR_TUPLE@498..511
                          LParen@498..499 "("
                          EXPR_BOOL@499..504
                            FalseKeyword@499..504 "false"
                          Comma@504..505 ","
                          Whitespace@505..506 " "
                          EXPR_BOOL@506..510
                            TrueKeyword@506..510 "true"
                          RParen@510..511 ")"
                        Comma@511..512 ","
                        Whitespace@512..513 " "
                      ARG@513..526
                        EXPR_TUPLE@513..526
                          LParen@513..514 "("
                          EXPR_BOOL@514..519
                            FalseKeyword@514..519 "false"
                          Comma@519..520 ","
                          Whitespace@520..521 " "
                          EXPR_BOOL@521..525
                            TrueKeyword@521..525 "true"
                          RParen@525..526 ")"
                      RParen@526..527 ")"
                      Whitespace@527..528 " "
                InKeyword@528..530 "in"
                Whitespace@530..533 "\n  "
                EXPR_LET_BODY@533..597
                  EXPR_LET@533..597
                    LetKeyword@533..536 "let"
                    Whitespace@536..537 " "
                    PATTERN_WILDCARD@537..539
                      WildcardKeyword@537..538 "_"
                      Whitespace@538..539 " "
                    Eq@539..540 "="
                    Whitespace@540..541 " "
                    EXPR_LET_VALUE@541..589
                      EXPR_CALL@541..589
                        EXPR_LIDENT@541..558
                          Lident@541..558 "test_nested_match"
                        ARG_LIST@558..589
                          LParen@558..559 "("
                          ARG@559..574
                            EXPR_TUPLE@559..572
                              LParen@559..560 "("
                              EXPR_BOOL@560..565
                                FalseKeyword@560..565 "false"
                              Comma@565..566 ","
                              Whitespace@566..567 " "
                              EXPR_BOOL@567..571
                                TrueKeyword@567..571 "true"
                              RParen@571..572 ")"
                            Comma@572..573 ","
                            Whitespace@573..574 " "
                          ARG@574..587
                            EXPR_TUPLE@574..587
                              LParen@574..575 "("
                              EXPR_BOOL@575..579
                                TrueKeyword@575..579 "true"
                              Comma@579..580 ","
                              Whitespace@580..581 " "
                              EXPR_BOOL@581..586
                                FalseKeyword@581..586 "false"
                              RParen@586..587 ")"
                          RParen@587..588 ")"
                          Whitespace@588..589 " "
                    InKeyword@589..591 "in"
                    Whitespace@591..594 "\n  "
                    EXPR_LET_BODY@594..597
                      EXPR_UNIT@594..597
                        LParen@594..595 "("
                        RParen@595..596 ")"
                        Whitespace@596..597 "\n"
      RBrace@597..598 "}"
      Whitespace@598..599 "\n"