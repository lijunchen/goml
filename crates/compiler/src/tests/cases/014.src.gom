type Tuple2_bool_bool struct {
    _0 bool
    _1 bool
}

func test_nested_match(x/0 Tuple2_bool_bool, y/1 Tuple2_bool_bool) {
    var x0 bool = x__0._0
    var x1 bool = x__0._1
    switch x1 {
    case true:
        var x2 bool = y__1._0
        var x3 bool = y__1._1
        switch x3 {
        case true:
            switch x2 {
            case true:
                return println("case4")
            case false:
                return println("case3")
            }
        case false:
            return println("case4")
        }
    case false:
        switch x0 {
        case true:
            var x4 bool = y__1._0
            var x5 bool = y__1._1
            switch x5 {
            case true:
                switch x4 {
                case true:
                    return println("case2")
                case false:
                    return println("case1")
                }
            case false:
                return println("case2")
            }
        case false:
            var x6 bool = y__1._0
            var x7 bool = y__1._1
            switch x7 {
            case true:
                switch x6 {
                case true:
                    return println("case4")
                case false:
                    return println("case3")
                }
            case false:
                return println("case4")
            }
        }
    }
}

func main() {
    var t12 = Tuple2_bool_bool{
        _0: true,
        _1: false,
    }
    var t13 = Tuple2_bool_bool{
        _0: false,
        _1: true,
    }
    var mtmp8 Unit = test_nested_match(t12, t13)
    var t14 = Tuple2_bool_bool{
        _0: true,
        _1: false,
    }
    var t15 = Tuple2_bool_bool{
        _0: true,
        _1: false,
    }
    var mtmp9 Unit = test_nested_match(t14, t15)
    var t16 = Tuple2_bool_bool{
        _0: false,
        _1: true,
    }
    var t17 = Tuple2_bool_bool{
        _0: false,
        _1: true,
    }
    var mtmp10 Unit = test_nested_match(t16, t17)
    var t18 = Tuple2_bool_bool{
        _0: false,
        _1: true,
    }
    var t19 = Tuple2_bool_bool{
        _0: true,
        _1: false,
    }
    var mtmp11 Unit = test_nested_match(t18, t19)
    return nil
}