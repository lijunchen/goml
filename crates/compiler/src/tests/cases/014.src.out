()
====
case1
case2
case3
case4

====
case1
case2
case3
case4
